### ISSUE TITLE
[Proposal] Foreign key shorthand on migrations

### ISSUE THREAD

**Issue by iget-master** [laravel/ideas#815]
Everytime that a user need to create a foreign key on migrations, he need to write this:

```
$table->unsignedInteger('user_id');
$table->foreign('user_id')->references('id')->on('users');
```

My proposal is to implement a shorthand to this like:

```
$table->unsignedInteger('user_id')->foreign();
```

On this shorthand, the migration try to guess the referenced table and column based on column name, as it's done on eloquent sometimes.

If user want to override the defaults, it can expand to the fluent way:

```
$table->unsignedInteger('user_id')->foreign()->references('id')->on('users');
```

I know that's a simple thing to implement, but can save a lot of time when you write large database structures, some of our projects have more then 100 tables with 1~3 foreign keys, and it's a pain to repeat so many code. :smile: 

If the proposal is accepted, our team can implement it and send a PR.

**Comment by dwightwatson** [comment]
I like this one. I end up having to look up the syntax for foreign keys each time i use them too, even though it reads as a sentence. It would be good if you could chain it on as `foreign()` and also call `$table->foreign('user_id')` like you can with other index methods as well.

**Comment by drbyte** [comment]
The fluent syntax mentioned is already available:

`$table->unsignedInteger('memberID')->references('id')->on('users')->index('byMember');`

I realize this doesn't auto-guess the relations, but it's simple and expressive, and specific, and doesn't require multiple lines/statements.

**Comment by iget-master** [comment]
@drbyte I didn't tried it, it's undocumented feature, right?

**Comment by drbyte** [comment]
Not sure if it's undocumented. But that's how I've been doing it since at least Laravel 5.0 -- I think I saw it done that way in a Laracasts lesson.

**Comment by iget-master** [comment]
@drbyte Nice, so I'll close this issue, since it's already exists.

**Comment by iget-master** [comment]
Actually @drbyte this doesn't work.

You example:

    $table->unsignedInteger('memberID')->references('id')->on('users')->index('byMember');

Just creates an common index, not a foreign key.

I've even tried:

    $table->unsignedInteger('memberID')->foreign()->references('id')->on('users');

but it just don't creates any indexes.

**Comment by drbyte** [comment]
You're right. 
I was wrong. It doesn't create the constraint as I understood it to.
Maybe it did at one time, but it definitely doesn't right now.

So I vote for it too.
However, it wouldn't surprise me that it's written the way it is because certain statements have to be run separately when executing the SQL inside the respective engines ... and the fluent syntax won't allow that.

**Comment by iget-master** [comment]
Yeah, I did relied on this feature on a project, and now noticed that no foreign keys exists. :-/

I'll take a time to read the code and propose some way to implement it.

**Comment by iget-master** [comment]
The Schema Blueprint class have a method called `morphs()` that is a shorthand to create polymorphic relationship fields. What do you guys think of a simillar method `belongs()` or `belongsTo()` with the following signature:

`public function belongsTo($table, $foreignKey = null, $ownerKey = null)`

That internally creates the field itself and the foreign key index?

I know that some databases doesn't have foreign key, but if this is the case, the user can use the `unsignedInt()` method only.

**Comment by sisve** [comment]
Todays migrations are a bit finicky. You can basically call anything you want on them.

    $table->unsignedInteger('memberID')->references('id')->on('users')->index('byMember')->shazaam('wee');

It builds up a Fluent object behind the scene and any method call is accepted. It's the up to the schema grammar to pick what options it understands, and ignore the rest. So in this case it becomes, assuming you're using Mysql, `MySqlGrammar::typeInteger`, `MySqlGrammar::modifyUnsigned`, and via the fluent index syntax `MySqlGrammar::compileIndex`.

All the other invocations just sets unused properties on the Fluent object. The problem here is that if we change this, existing code may go bonkers. Sure, in the absolute best case scenario we start doing what everyone has always wanted but forgot to verify; but there's a high probability that we'll forgot about some weird things people have written, that already works (as in; not crashes) that will stop working (crashes).

Maybe we could have a release that, during development, throws warnings for unsupported modifiers (in my example: references, on, shazaam) so that people can fix their migrations or at least be aware their migrations has unsupported stuff ... for the current grammar. Since the grammar picks what it wants to understand, anyone having a migration that executes against several databases engines will have to use only the features that every database engine supports. That means; no more unsigned numerics (affects our auto_increment keys, but unsigned numerics isn't part of the sql standard anyway), charset/collate, comments, or after/first for positioning new columns.

Or we could implement those features in all grammars.

**TL;DR:** Introducing this has the potential to cause huge problems for existing migrations that have code that until this has been ignored.

**Comment by drbyte** [comment]
Thanks @sisve - I was looking forward to your input!

I agree: given that unsupported syntax is silently ignored, it actually makes migrations more brittle. To your point they may produce unexpected results if changing database engines.

Do you think a PR which offers a `php artisan migrate:inspect` option would be welcomed if its purpose was to point out unsupported uses on the current db engine? and maybe secondarily to point out usages which stray from the SQL standard?

**Comment by sisve** [comment]
A command to verify the migrations sounds reasonable, but very hard to write. We cannot just execute migrations and hope they have no side-effects and that there's no weird calls directly to the DB facade or pdo instances. So, if we cannot reliably execute the migrations to get the data for the inspection, how would such a command be implemented?

There's some validation of these things in Doctrine when writing the mappings; you could use the xml-schema to verify that your xml mappings are correct. (The xml schema has other validation related issues instead.) However, there's nothing I know of that will look at php code and verify that it is correct.

The only thing I can think of with todays way of writing migration system would be for our current migration system to throw NotSupportedException when it detects these unknown column methods. This would change the current behavior from "silently ignoring" to "frantically screaming". This, of course, means we have code that goes from working to not working. Perhaps we could throw the exception when app.debug is enabled, and silently ignore the problem in production?

We could stop using the Fluent class to back up the column definitions, and instead an actual typed Column. We could either write methods on it and only allow those, or we could extend the Fluent and add the supported methods as `@method` annotations. This way we would at least get ide support to highlight unknown methods.

**Comment by taylorotwell** [comment]
No plans on changing this for now.

**Comment by iget-esoares** [comment]
@taylorotwell What about adding a `belongs` method just like the `morphs` one? I can do a PR.
